### ISSUE TITLE
Faster units`

### ISSUE THREAD

**Issue by zchen088** [labrad/pylabrad#213]
@ejeffrey, you mentioned a while back that you had code to make unit computations 100x faster. Is there any chance we could get that deployed?

**Comment by ejeffrey** [comment]
Yes.  The code as it currently stands is in u/ejeffrey/fast_units under pylabrad/labrad/fast_units.py and pylabrad/labrad/unit_array.c.  There are a couple of caveats that need to be addressed before we deploy it:

1) Hasn't been tested against the existing python version. 
2) The module contains C code that has to be compiled.  So we need to figure out how to deploy it in the least painful fashion possible
3) It is not exactly API compatible.  The main issue is that  right now there is only a single 'Value' class, rather than separate Value/Complex/ValueArray/DimensionlessValue classes.  This should be fixed.  There is also no 'WithUnit' base class.  This is fairly easy to fix.

**Comment by DanielSank** [comment]
> 2) The module contains C code that has to be compiled. So we need to figure out how to deploy it in the least painful fashion possible

One possibility (for which I vote) is to make this library external to pylabrad and put it on PyPI. Then we can make wheels etc. and nobody has to do anything tricky to install it, including Windows users.

**Comment by maffoo** [comment]
@DanielSank, pylabrad wheels are already on PyPI, so that in itself is not an argument for splitting this out from pylabrad (the current wheels have no C extensions and so are platform independent, but still). That being said, I like the idea of having this unit library be external to pylabrad proper. The question would then be how to make it possible to use the library with pylabrad. I'd prefer to clean up pylabrad so that it's possible to plug in flatteners/unflatteners for custom types, but we have to define how that would work.

**Comment by DanielSank** [comment]
> pylabrad wheels are already on PyPI, so that in itself is not an argument for splitting this out from pylabrad

Well, given that everyone currently has a "dev" version of pylabrad checked out in their file system and has their pythonpath including that, I do think having the units library separated out as an installed lib is different, but whatever.

**Comment by ejeffrey** [comment]
To update this, I have been working on a high performance unit implementation written in C.  You can look at the code in u/ejeffrey/c_units.  It is largely compatible with the existing Value/Unit classes as long, but it has a few differences:

* zero is not special cased, comparisons and addition to float(0) will cause an error unless the other argument is dimensionless. (see #140)
* Units with offsets are not supported at all (see #141)
* DimensionlessValue, DimensionlessComplex, and DimensionlessArray classes do not exist.  Previously, these classes implemented the WithUnit API, but subclassed float, complex and ndarray.  Instead, WithUnits objects implement __float__, __complex__, and __array__ methods.  In almost but (maybe) not all cases that expect a number or array object, these functions will automatically be called to do the conversion.

**Comment by jayich** [comment]
@ejeffrey any updates on fast units?
